File: src/main/java/bot/farm/snb/repository/UserGameStateRepository.java
package bot.farm.snb.repository;

import bot.farm.snb.entity.Game;
import bot.farm.snb.entity.User;
import bot.farm.snb.entity.UserGameState;
import java.util.List;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

/**
 * Repository interface for managing UserGameState entities.
 */
@Repository
public interface UserGameStateRepository extends JpaRepository<UserGameState, Long> {
  /**
   * Finds a UserGameState by user and game.
   *
   * @param user The User entity.
   * @param game The Game entity.
   * @return The UserGameState entity.
   */
  UserGameState findByUserAndGame(User user, Game game);

  /**
   * Finds a UserGameState by user's chat ID and game's name.
   *
   * @param chatId The user's chat ID.
   * @param name   The game's name.
   * @return The UserGameState entity.
   */

  UserGameState findByUser_ChatIdAndGame_Name(String chatId, String name);

  /**
   * Finds a list of UserGameStates by user's chat ID and isBanned is true.
   *
   * @param chatId The user's chat ID.
   * @return The list of UserGameState entities.
   */
  List<UserGameState> findByUser_ChatIdAndIsBannedTrue(String chatId);

  /**
   * Updates the isBanned field of a UserGameState by ID.
   *
   * @param isBanned The new value for isBanned.
   * @param id       The ID of the UserGameState entity.
   */
  @Transactional
  @Modifying
  @Query("update UserGameState u set u.isBanned = ?1 where u.id = ?2")
  void updateIsBannedById(boolean isBanned, Long id);

  /**
   * Updates the isWished and isOwned fields of a UserGameState by ID.
   *
   * @param isWished The new value for isWished.
   * @param isOwned  The new value for isOwned.
   * @param id       The ID of the UserGameState entity.
   */
  @Transactional
  @Modifying
  @Query("update UserGameState u set u.isWished = ?1, u.isOwned = ?2 where u.id = ?3")
  void updateIsWishedAndIsOwnedById(boolean isWished, boolean isOwned, Long id);

  /**
   * Deletes UserGameStates associated with a user.
   *
   * @param user The User entity.
   */
  @Transactional
  @Modifying
  @Query("delete from UserGameState u where u.user = ?1")
  void deleteByUser(User user);

  /**
   * Checks if a UserGameState exists for a user and game.
   *
   * @param user The User entity.
   * @param game The Game entity.
   * @return {@code true} if the UserGameState exists, {@code false} otherwise.
   */
  boolean existsByUserAndGame(User user, Game game);

  /**
   * Finds the top games based on the count of UserGameStates.
   *
   * @param limit The maximum number of games to retrieve.
   * @return The list of game names.
   */
  @Query(value = "SELECT g.name FROM user_game_state ugs "
      + "JOIN games g ON ugs.game_id = g.game_id "
      + "GROUP BY ugs.game_id, g.name "
      + "HAVING COUNT(ugs.game_id) > 5 "
      + "ORDER BY COUNT(ugs.game_id) "
      + "DESC LIMIT ?1", nativeQuery = true)
  List<String> findTopGames(int limit);
}

File: steam_news_bot_kt/src/main/kotlin/repository/UserGameStateRepository.kt
package sidim.doma.repository

import org.jetbrains.exposed.sql.*
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import org.jetbrains.exposed.sql.transactions.transaction
import sidim.doma.entity.Games
import sidim.doma.entity.UserGameState
import sidim.doma.entity.UserGameStates

class UserGameStateRepository {
    fun createUgs(ugs: List<UserGameState>) {
        return transaction {
            UserGameStates.batchInsert(ugs) {
                this[UserGameStates.userId] = it.userId
                this[UserGameStates.gameId] = it.gameId
                this[UserGameStates.isWished] = it.isWished
                this[UserGameStates.isBanned] = it.isBanned
                this[UserGameStates.isOwned] = it.isOwned
            }
        }
    }

    fun deleteUgsByUserId(userId: String) {
        return transaction {
            UserGameStates
                .deleteWhere { UserGameStates.userId eq userId }
        }
    }

    fun findByUserIdAndGameId(userId: String, gameId: String): UserGameState? {
        return transaction {
            UserGameStates
                .selectAll()
                .where {
                    (UserGameStates.userId eq userId) and
                            (UserGameStates.gameId eq gameId)
                }
                .singleOrNull()
                ?.let { rowToUserGameState(it) }
        }
    }

    fun findByUserIdAndGameName(userId: String, gameName: String): UserGameState? {
        return transaction {
            (UserGameStates innerJoin Games)
                .selectAll()
                .where {
                    (UserGameStates.userId eq userId) and
                            (Games.name eq gameName)
                }
                .singleOrNull()
                ?.let { rowToUserGameState(it) }
        }
    }

    fun findByUserIdAndBanned(userId: String, isBanned: Boolean): UserGameState? {
        return transaction {
            UserGameStates
                .selectAll()
                .where {
                    (UserGameStates.userId eq userId) and
                            (UserGameStates.isBanned eq isBanned)
                }
                .singleOrNull()
                ?.let { rowToUserGameState(it) }
        }
    }

    fun updateIsBannedById(isBanned: Boolean, id: Long) {
        transaction {
            UserGameStates.update({ UserGameStates.id eq id }) {
                it[UserGameStates.isBanned] = isBanned
            }
        }
    }

    fun updateIsBannedByGameNameAndUserId(banned: Boolean, gameName: String, userId: String) {
        transaction {
            (UserGameStates innerJoin Games)
                .update({ (UserGameStates.userId eq userId) and (Games.name eq gameName) }) {
                    it[UserGameStates.isBanned] = banned
                }
        }
    }

    fun updateIsWishedAndIsOwnedById(isWished: Boolean, isOwned: Boolean, id: Long) {
        transaction {
            UserGameStates.update({ UserGameStates.id eq id }) {
                it[UserGameStates.isWished] = isWished
                it[UserGameStates.isOwned] = isOwned
            }
        }
    }

    fun existsByUserIdAndGameId(userId: String, gameId: String): Boolean {
        return transaction {
            UserGameStates
                .selectAll()
                .where {
                    (UserGameStates.userId eq userId) and
                            (UserGameStates.gameId eq gameId)
                }
                .singleOrNull()
                ?.let { true } ?: false
        }
    }

    fun countByUserIdAndIsOwned(userId: String, isOwned: Boolean): Long {
        return transaction {
            UserGameStates
                .selectAll()
                .where {
                    (UserGameStates.userId eq userId) and
                            (UserGameStates.isOwned eq isOwned)
                }
                .count()
        }
    }

    fun countByUserIdAndIsWished(userId: String, isWished: Boolean): Long {
        return transaction {
            UserGameStates
                .selectAll()
                .where {
                    (UserGameStates.userId eq userId) and
                            (UserGameStates.isWished eq isWished)
                }
                .count()
        }
    }

    private fun rowToUserGameState(it: ResultRow) = UserGameState(
        userId = it[UserGameStates.userId],
        gameId = it[UserGameStates.gameId],
        isWished = it[UserGameStates.isWished],
        isBanned = it[UserGameStates.isBanned],
        isOwned = it[UserGameStates.isOwned]
    )

    fun clearBlackListByUserId(userId: String) {
        return transaction {
            UserGameStates.update({
                (UserGameStates.userId eq userId) and (UserGameStates.isBanned eq true)
            }) {
                it[isBanned] = false
            }
        }
    }
}